src/itsdangerous/serializer.py

```diff
@@ -10,13 +10,20 @@
 from .signer import _make_keys_list
 from .signer import Signer
 
+_TAnyStr = t.TypeVar("_TAnyStr", str, bytes, covariant=True)
 
-def is_text_serializer(serializer: t.Any) -> bool:
+
+class _PDataSerializer(t.Protocol[_TAnyStr]):
+    def loads(self, payload: str | bytes) -> t.Any: ...
+    def dumps(self, obj: t.Any, **kwargs: t.Any) -> _TAnyStr: ...
```

Comment:
```suggestion
class _PDataSerializer(t.Protocol[t.AnyStr]):
    def loads(self, payload: t.AnyStr, /) -> t.Any: ...
    def dumps(self, obj: object, /) -> t.AnyStr: ...
```
You are providing a lower bound here, so you don't want to make the argument list too strict.

For `loads` you can't use `str | bytes` because parameters are contravariant, so that would require each serializer to accept both `str` and `bytes`. Exactly because parameters are contravariant you can satisfy a function that expects `str` with a supertype of `str`, such as a union with `bytes`.

As for `dumps`, IIRC unless you specifically use `*args: Any, **kwargs: Any` neither mypy nor pyright will treat this as a gradual signature, that allows serializers with arbitrary extra arguments, it will instead require each serializer function to take arbitrary arguments. This means you will have to ignore a type error when you pass in `serializer_kwargs`, but that's a hit you will have to take until we have something like a `TypeVarMapping` to pass just `kwargs` along, `ParamSpec` doesn't really work here.

Reply:
Yep, after making these changes mypy does accept `json` and `pickle` as valid.

Reply:
pyright is still not happy though.